Implement Deploy buffer tests

alsrdn opened the pull request:
Implements tests for the deploy buffer.
Also fix some issues encountered while implementing tests.

Fixes: https://github.com/casper-network/casper-node/issues/3447

alsrdn committed:
AppendableBlock: fix potential panic when invalid config is used

When an invalid `DeployConfig` is used, the implementation of
`AppendableBlock::would_exceed_approval_limits` can panic due to
arithmetic operations that can overflow. Replace these with safer
variants. Also add some tests for the use case that was failing.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
deploy_buffer: fix block continuity check when validating full ttl
awareness

When checking if we have full ttl worth of deploys in order to detect
deploy replays, we also check for gaps in the chain index that we have.
Because this check was made after we check the timestamp, it may be
possible to miss a discontinuity if the previous block is not at the
propper height but has a timestamp older than the ttl.
Fix this issue by moving the timestamp check after the block continuity
check.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
tests: add more utility functions for generating deploys and blocks

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
deploy_buffer: add tests

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

EdHastingsCasperAssociation commented on node/src/components/deploy_buffer/tests.rs:
```diff
@@ -0,0 +1,590 @@
+use rand::Rng;
+
+enum DeployType {
+    NativeTransfer,
```
Transfer

alsrdn replied:
changed here: https://github.com/casper-network/casper-node/pull/3544/commits/3235906ceeb460ac3160bb23ccff0125ed409184

EdHastingsCasperAssociation commented on node/src/components/deploy_buffer/tests.rs:
```diff
@@ -0,0 +1,590 @@
+
+enum DeployType {
+    NativeTransfer,
+    Contract,
```
Standard

EdHastingsCasperAssociation commented on node/src/components/deploy_buffer.rs:
```diff
@@ -146,6 +143,12 @@ impl DeployBuffer {
                 debug!("DeployBuffer: missing block at height: {}", contiguous_next);
                 return false;
             }
+
+            // if we've reached genesis via an unbroken sequence, we're good
+            // if we've seen a full ttl period of blocks, we're good
+            if *height == 0 || *timestamp < earliest_needed {
+                return true;
+            }
```
curious why you pulled this lower?

alsrdn replied:
Because it would skip the unbroken sequence check if we reach the genesis block or the right timestamp. For example if we had blocks [0, 2, 3, 4, 5] it would still return true even though block 1 was missing.

georgepisaltu commented on node/src/components/deploy_buffer/tests.rs:
```diff
@@ -0,0 +1,590 @@
+}
+
+#[test]
+fn test_register_deploy() {
```
For consistent naming, we shouldn't have test names beginning with "test". Instead, test names usually start with either the component name followed by the function it's testing or the function it's testing followed by the expected action.

E.g. "block_synchronizer_purge_historical", "need_next_invalid_state_should_fail"

Applies to all tests in this PR.

alsrdn replied:
changed here: https://github.com/casper-network/casper-node/pull/3544/commits/3235906ceeb460ac3160bb23ccff0125ed409184

georgepisaltu reviewed (commented):
LGTM overall, just one comment.

alsrdn committed:
tests/deploy_buffer: rename tests for consistency

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
tests/deploy_buffer: fix lint errors

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

fizyk20 commented on node/src/types/appendable_block.rs:
```diff
@@ -190,7 +190,12 @@ impl AppendableBlock {
             + self.deploy_config.block_max_deploy_count as usize
             - self.deploys.len();
-        additional_approvals > remaining_approval_slots - remaining_deploy_slots + 1
+        if let Some(total_approvals) = additional_approvals.checked_add(remaining_deploy_slots) {
```
I think calling it `total_approvals` might be confusing, since this is a sum of a number of actual approvals and a number representing deploy slots... I'd call it something like `approvals_plus_deploy_slots` instead.

I don't really understand why we need this change, though - we should never hit the case of `remaining_deploy_slots` being greater than `remaining_approval_slots` precisely because of this check.

alsrdn replied:
After a discussion with @fizyk20 we agreed that it's better to actually add a validation of the deploy config parameters and revert to the previous implementation. See https://github.com/casper-network/casper-node/pull/3544/commits/bb3b3394119146f7937711109d237ff6e018fc31 and https://github.com/casper-network/casper-node/pull/3544/commits/fc3597e3db831f57ef80d08c925797542c9f70a7

fizyk20 reviewed (approved):
Looks good, just one non-blocking nitpick.

alsrdn committed:
chainspec: add deploy config validation

Validate that the max number of deploys + transfers in a block is not
allowed to be larger than the max number of approvals.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
Partially revert "AppendableBlock: fix potential panic when invalid config is used"

This partially reverts commit a821b4f145c15f80ced13b0fbd4b9fba313377a7.
The changes are not needed since now we validate the deploy config
at chainspec load time.
Added a comment to illustrate this. Also kept test utility function that
allows access to the deploy and transfer set.

rafal-ch reviewed (approved):
LGTM :+1:

alsrdn commented:
bors try

alsrdn commented:
bors merge
